core/src/OpenACC/Kokkos_OpenACC_Atomic.hpp

```diff
@@ -0,0 +1,1964 @@
+#else
+  unsigned int *ptr = const_cast<unsigned int *>(dest);
+#pragma acc atomic write
+    ptr[0] = val;
+#endif
+}
+
+#pragma acc routine seq
+template <>
+ __inline__ void atomic_assign(volatile unsigned long long int *const dest, const unsigned long long int &val) {
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	atomicExch((unsigned long long int *)dest, val);
+#else
+  unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+#pragma acc atomic write
+    ptr[0] = val;
+#endif
+}
+
+// // atomic_fetch_add ------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+inline  T atomic_fetch_add(
+    volatile T *const dest,
+    const T &val) {
+    T retval;
+	T *ptr = const_cast<T *>(dest);
+//#pragma acc atomic capture
+    {
+      retval = ptr[0];
+      ptr[0] += val;
+    }
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+inline  int atomic_fetch_add(
+    volatile int *const dest,
+    const int &val) {
+    int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicAdd((int *)dest, val);
+#else
+	int *ptr = const_cast<int *>(dest);
+#pragma acc atomic capture
+    {
+      retval = ptr[0];
+      ptr[0] += val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+inline  unsigned int atomic_fetch_add(
+    volatile unsigned int *const dest,
+    const unsigned int &val) {
+    unsigned int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicAdd((unsigned int *)dest, val);
+#else
+	unsigned int *ptr = const_cast<unsigned int *>(dest);
+#pragma acc atomic capture
+    {
+      retval = ptr[0];
+      ptr[0] += val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+inline  unsigned long long int atomic_fetch_add(
+    volatile unsigned long long int *const dest,
+    const unsigned long long int &val) {
+    unsigned long long int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicAdd((unsigned long long int *)dest, val);
+#else
+	unsigned long long *ptr = const_cast<unsigned long long *>(dest);
+#pragma acc atomic capture
+    {
+      retval = ptr[0];
+      ptr[0] += val;
+    }
+#endif
+    return retval;
+}
+
+/*
+#pragma acc routine seq
+template <>
+inline  int64_t atomic_fetch_add(
+    volatile int64_t *const dest,
+    const int64_t &val) {
+    int64_t retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicAdd((int64_t *)dest, val);
+#else
+	int64_t *ptr = const_cast<int64_t *>(dest);
+#pragma acc atomic capture
+    {
+      retval = ptr[0];
+      ptr[0] += val;
+    }
+#endif
+    return retval;
+}
+*/
+
+// atmic_fetch_sub -------------------------------------------------------------
+
+#pragma acc routine seq
+template <class T>
+ __inline__ T atomic_fetch_sub(
+    volatile T * const dest,
+    const T & val) {
+    T retval;
+	T *ptr = const_cast<T *>(dest);
+//#pragma acc atomic capture
```

Comment:
That code was for unsupported atomic data types; I changed the code so that for unsupported atomic data types, it will abort if running on a device but it will sequentially execute the construct (without atomicity) if running on a host.